File: src/com/goodtrendltd/HolySongs/DisplayLyricActivity.java
package com.goodtrendltd.HolySongs;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.provider.Settings;
import android.text.method.ScrollingMovementMethod;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.widget.ShareActionProvider;
import android.widget.TextView;

/**
 * Created with IntelliJ IDEA.
 * User: LeOn
 * Date: 13-10-5
 * Time: 下午10:13
 */
public class DisplayLyricActivity extends Activity {
    public static String SEARCH_TARGET = "com.goodtrendltd.searchTarget";

    private ShareActionProvider mShareActionProvider;
    private String songName;
    private String lyric;

    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        SharedPreferences sharedPreferences = getSharedPreferences(getString(R.string.app_pref), MODE_PRIVATE);
        if (sharedPreferences.getBoolean(getString(R.string.night_mode_pref_key), true)) {
            setTheme(R.style.MyHoloTheme);
        } else {
            setTheme(R.style.MyHoloLightTheme);
        }
        setContentView(R.layout.lyric_view);

        Intent intent = getIntent();
        lyric = intent.getStringExtra(MainActivity.LYRIC);
        songName = intent.getStringExtra(MainActivity.SONG_NAME);

        // Create the text view
        TextView lyricTextView = (TextView) findViewById(R.id.lyricContent);
        lyricTextView.setMovementMethod(new ScrollingMovementMethod());
        lyricTextView.setTextSize(sharedPreferences.getInt(getString(R.string.font_size_pref_key), 20));
        lyricTextView.setText(lyric);
        lyricTextView.setLineSpacing(20, 1.0f);

        TextView titleTextView = (TextView) findViewById(R.id.title);
        titleTextView.setText(songName);
        titleTextView.setTextSize(25);
        titleTextView.setTextColor(Color.parseColor("#00BFFF"));

    }

    //begin of menu related
    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu items for use in the action bar
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.lyricview_actions, menu);
        // Locate MenuItem with ShareActionProvider
        MenuItem item = menu.findItem(R.id.menu_item_share);
        // Fetch and store ShareActionProvider
        mShareActionProvider = (ShareActionProvider) item.getActionProvider();

        Intent i = new Intent(Intent.ACTION_SEND);
        i.setType("text/plain");
        i.putExtra(Intent.EXTRA_SUBJECT, songName);
        i.putExtra(Intent.EXTRA_TEXT, lyric);

        mShareActionProvider.setShareIntent(i);

        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int itemId = item.getItemId();
        if (itemId == R.id.youtube) {
            openSearch(getString(R.string.youtube));
            return true;
        }
        if (itemId == R.id.youku) {
            openSearch(getString(R.string.youku));
            return true;
        }
        if (itemId == R.id.tudou) {
            openSearch(getString(R.string.tudou));
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void openSearch(final String target) {
        ConnectivityManager manager = (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo.State wifi = manager.getNetworkInfo(ConnectivityManager.TYPE_WIFI).getState();
        if (wifi == NetworkInfo.State.CONNECTED || wifi == NetworkInfo.State.CONNECTING) {
            navigateToSearch(target);
        } else {
            AlertDialog.Builder builder = new AlertDialog.Builder(this);
            builder.setIcon(android.R.drawable.ic_dialog_alert);
            builder.setTitle(R.string.alert_title);
            builder.setMessage(R.string.alert_message);
            builder.setPositiveButton(R.string.ok_button, new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    // 如果没有网络连接，则进入网络设置界面
                    navigateToSearch(target);
                }
            });
            builder.setNegativeButton(R.string.cancel_button, new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    dialog.cancel();
                }
            });
            builder.create();
            builder.show();
        }
    }

    private void navigateToSearch(String target){
        Intent intent = new Intent(this, VideoSearch.class);
        intent.putExtra(SEARCH_TARGET, target);
        intent.putExtra(MainActivity.SONG_NAME, songName);
        startActivity(intent);
    }

    // Call to update the share intent
    private void setShareIntent(Intent shareIntent) {
        if (mShareActionProvider != null) {
            mShareActionProvider.setShareIntent(shareIntent);
        }
    }
}

File: app/src/main/kotlin/com/goodtrendltd/HolySongs/DisplayLyricActivity.kt
package com.goodtrendltd.HolySongs

import android.app.AlertDialog
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.goodtrendltd.HolySongs.data.ReaderPreferenceSnapshot
import com.goodtrendltd.HolySongs.data.ReaderPreferences
import com.goodtrendltd.HolySongs.ui.HolySongsTheme
import com.goodtrendltd.HolySongs.ui.LyricScreen

/** Compose lyric host retaining the original String extras and provider entry contract. */
class DisplayLyricActivity : ComponentActivity() {
    companion object {
        @JvmField
        val SEARCH_TARGET: String = "com.goodtrendltd.searchTarget"

        @JvmStatic
        fun validStringExtra(extras: Bundle?, key: String?): String? {
            if (extras == null || key == null || !extras.containsKey(key)) return null
            val value = extras.get(key)
            return value as? String
        }
    }

    private val readerPreferences by lazy { ReaderPreferences(applicationContext) }
    private var songName: String = ""
    private var lyric: String = ""
    // Activity-owned one-shot guards prevent recomposition/rapid taps from duplicating effects.
    private val launchGuards = DisplayLyricLaunchGuards()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        val extras = intent?.extras
        lyric = validStringExtra(extras, MainActivity.LYRIC) ?: return finishInvalid()
        songName = validStringExtra(extras, MainActivity.SONG_NAME) ?: return finishInvalid()
        if (songName.isEmpty()) return finishInvalid()

        setContent {
            val preferences = readerPreferences.snapshots.collectAsStateWithLifecycle(
                initialValue = ReaderPreferenceSnapshot(
                    fontSize = 20,
                    effectiveFontSize = 20,
                    nightMode = true,
                ),
            ).value
            HolySongsTheme(preferences.nightMode) {
                LyricScreen(
                    title = songName,
                    lyric = lyric,
                    preferences = preferences,
                    onShare = ::shareLyric,
                    onVideo = ::openSearch,
                )
            }
        }
    }

    private fun finishInvalid() {
        finish()
    }

    private fun shareLyric() {
        if (!launchGuards.tryStartShare()) return
        val send = Intent(Intent.ACTION_SEND).apply {
            type = "text/plain"
            putExtra(Intent.EXTRA_SUBJECT, songName)
            putExtra(Intent.EXTRA_TEXT, lyric)
        }
        try {
            startActivity(Intent.createChooser(send, "选择用于分享的app"))
        } catch (_: ActivityNotFoundException) {
            launchGuards.resetShare()
            // A device with no recipient remains on the offline lyric screen.
        } catch (_: RuntimeException) {
            launchGuards.resetShare()
            // A broken chooser/window also leaves the offline lyric screen usable.
        }
    }

    private fun openSearch(target: String) {
        if (!isSupportedTarget(target) || songName.isEmpty() || !launchGuards.tryStartVideo()) return
        val manager = getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
        val network = manager?.activeNetwork
        if (network == null) {
            launchGuards.resetVideo()
            return
        }
        val capabilities = manager.getNetworkCapabilities(network)
        if (capabilities?.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) == true) {
            navigateToSearch(target)
        } else {
            showMobileDataConfirmation(target)
        }
    }

    private fun showMobileDataConfirmation(target: String) {
        AlertDialog.Builder(this)
            .setIcon(android.R.drawable.ic_dialog_alert)
            .setTitle(R.string.alert_title)
            .setMessage(R.string.alert_message)
            .setPositiveButton(R.string.ok_button) { _, _ -> navigateToSearch(target) }
            .setNegativeButton(R.string.cancel_button) { dialog, _ -> dialog.cancel() }
            .setOnCancelListener { launchGuards.resetVideo() }
            .show()
    }

    private fun navigateToSearch(target: String) {
        if (!isSupportedTarget(target) || songName.isEmpty()) {
            launchGuards.resetVideo()
            return
        }
        val videoIntent = Intent(this, VideoSearch::class.java)
            .putExtra(SEARCH_TARGET, target)
            .putExtra(MainActivity.SONG_NAME, songName)
        try {
            startActivity(videoIntent)
        } catch (_: ActivityNotFoundException) {
            launchGuards.resetVideo()
            // Keep the lyric visible if the retained provider boundary is unavailable.
        } catch (_: RuntimeException) {
            launchGuards.resetVideo()
            // Keep the lyric visible if the retained provider boundary cannot be launched.
        }
    }

    override fun onResume() {
        super.onResume()
        // Returning from a chooser/video (or a failed transient window) permits a later action.
        launchGuards.resetOnReturn()
    }

    private fun isSupportedTarget(target: String): Boolean =
        target == getString(R.string.youtube) ||
            target == getString(R.string.douyin)
}
